sdk/eventgrid/Azure.Messaging.EventGrid/EventGridSourceGenerator/src/EventGridSourceGenerator.cs

```diff
@@ -13,38 +18,168 @@ namespace Azure.EventGrid.Messaging.SourceGeneration
     /// from constant values to deserialization method for each system event.
     /// </summary>
     [Generator]
-    internal class EventGridSourceGenerator : ISourceGenerator
+    internal class EventGridSourceGenerator : IIncrementalGenerator
     {
-        private SourceVisitor _visitor;
-        private bool _isSystemEventsLibrary;
         private const string Indent = "    ";
 
-        public void Execute(GeneratorExecutionContext context)
+        // the event name is either 3 or 4 parts, e.g. Microsoft.AppConfiguration.KeyValueDeleted or Microsoft.ResourceNotifications.HealthResources.AvailabilityStatusChanged
+        private static readonly Regex EventTypeRegex = new("[a-zA-Z]+\\.[a-zA-Z]+\\.[a-zA-Z]+(\\.[a-zA-Z]+)?", RegexOptions.Compiled);
+
+        private static ReadOnlySpan<char> SummaryStartTag => "<summary>".AsSpan();
+        private static ReadOnlySpan<char> SummaryEndTag => "</summary>".AsSpan();
+
+        public void Initialize(IncrementalGeneratorInitializationContext context)
         {
-            _visitor = new SourceVisitor();
-            _isSystemEventsLibrary = context.Compilation.AssemblyName == "Azure.Messaging.EventGrid.SystemEvents";
-            var root = context.Compilation.GetSymbolsWithName(
-                "SystemEvents",
-                SymbolFilter.Namespace)
-                .Single();
-            _visitor.Visit(root);
-
-            context.AddSource("SystemEventNames.cs", SourceText.From(ConstructSystemEventNames(), Encoding.UTF8));
-            context.AddSource("SystemEventExtensions.cs", SourceText.From(ConstructSystemEventExtensions(), Encoding.UTF8));
+            // Get all class declarations that end with "EventData"
+            var classDeclarations = context.SyntaxProvider
+                .CreateSyntaxProvider(
+                    predicate: static (s, _) => s is ClassDeclarationSyntax cds && cds.Identifier.Text.EndsWith("EventData"),
+                    transform: static (ctx, cancellationToken) =>
+                    {
+                        var semanticModel = ctx.SemanticModel;
+                        var classDecl = (ClassDeclarationSyntax)ctx.Node;
+
+                        // Get the symbol for the class declaration
+                        var symbol = semanticModel.GetDeclaredSymbol(classDecl, cancellationToken);
+                        if (symbol == null)
+                        {
+                            return null;
+                        }
+
+                        // Check if the symbol is in the SystemEvents namespace
+                        var containingNamespace = symbol.ContainingNamespace;
+                        while (containingNamespace != null)
```

Comment:
Given SystemEvents will always be the leaf do we need the while loop here?  Shouldn't we be able to inspect just one layer of ContainingNamespace?